File: code/tree.m
function[leaf] = make_leaf(node)
    leaf = cell(2,1);
    leaf{1} = 'leaf';
    leaf{2} = node;
endfunction

function[node] = get_values(leaf)
    node = leaf{2};
endfunction

function[val] = is_leaf(node)
    val = node{1} == 'leaf';
endfunction

function[node] = make_node(question, left, right)
    node = cell(2,1);
    node{1} = 'node';
    node{2} = question;
    children = struct();
    children.left = left;
    children.right = right;
    node{3} = children;
endfunction

function[val] = is_node(node)
    val = node{1} == 'node';
endfunction

function[question] = get_left(node)
    question = node{3}.left;
endfunction

function[question] = get_right(node)
    question = node{3}.right;
endfunction

function[question] = get_question(node)
    question = node{2};
endfunction

File: code/tree.py
import numpy as np

class Node:
    def __init__(self, question, left, right, left_points, right_points):
        self.question = question
        self.left = left
        self.right = right
	self.left_points = left_points
	self.right_points = right_points

    def get_question(self):
        return self.question

    def get_left(self):
        return self.left

    def get_right(self):
        return self.right

class Leaf:
    def __init__(self, collection):
        self.collection = collection

    def get_collection(self):
        return self.collection

class Question:
    def __init__(self, feature, value):
        self.feature = feature
        self.value = value

    def get_feature(self):
        return self.feature

    def get_value(self):
        return self.value

    def __repr__(self):
	return 'Question: [%d] <= %f' % (self.feature, self.value)
